runtime/runtime_metrics.go

```diff
@@ -0,0 +1,267 @@
+
+	// General
+	alloc      tally.Gauge // bytes of allocated heap objects
+	totalAlloc tally.Gauge // cumulative bytes allocated for heap objects
+	sys        tally.Gauge // total bytes of memory obtained from the OS
+	lookups    tally.Gauge // number of pointer lookups performed
+	mallocs    tally.Gauge // cumulative count of heap objects allocated
+	frees      tally.Gauge // cumulative count of heap objects freed
+
+	// Heap
+	heapAlloc    tally.Gauge // bytes of allocated heap objects
+	heapSys      tally.Gauge // bytes of heap memory obtained from the OS
+	heapIdle     tally.Gauge // bytes in idle (unused) spans
+	heapInuse    tally.Gauge // bytes in in-use spans
+	heapReleased tally.Gauge // bytes of physical memory returned to the OS
+	heapObjects  tally.Gauge // number of allocated heap objects
+
+	// Stack
+	stackInuse  tally.Gauge // bytes in stack spans
+	stackSys    tally.Gauge // bytes of stack memory obtained from the OS
+	mspanInuse  tally.Gauge // bytes of allocated mspan structures
+	mspanSys    tally.Gauge // bytes of memory obtained from the OS for mspan
+	mcacheInuse tally.Gauge // bytes of allocated mcache structures
+	mcacheSys   tally.Gauge // bytes of memory obtained from the OS for mcache structures
+
+	otherSys tally.Gauge // bytes of memory in miscellaneous off-heap runtime allocations
+
+	// GC
+	gcSys         tally.Gauge // bytes of memory in garbage collection metadata
+	nextGC        tally.Gauge // target heap size of the next GC cycle
+	lastGC        tally.Gauge // time the last garbage collection finished, as nanoseconds since epoch
+	pauseTotalMs  tally.Gauge // cumulative nanoseconds in GC stop-the-world pauses since the program started
+	pauseMs       tally.Gauge // most recent pause timing
+	numGC         tally.Gauge // number of completed GC cycles
+	gcCPUFraction tally.Gauge // fraction of cpu consumed by GC
+}
+
+// runtimeCollector keeps the current state of runtime metrics
+type runtimeCollector struct {
+	config       RuntimeMetricsConfiguration
+	scope        tally.Scope
+	metrics      runtimeMetrics
+	startedMutex sync.RWMutex
+	started      bool // protected by startedMutex
+	stop         chan struct{}
+}
+
+// StartRuntimeMetricsCollector starts collecting runtime metrics periodically.
+// Recommended usage:
+//     rm := StartRuntimeMetricsCollector(rootScope.Scope("runtime"), config)
+//     ...
+//     rm.Close()
+func StartRuntimeMetricsCollector(
+	config RuntimeMetricsConfiguration,
+	scope tally.Scope,
+) RuntimeMetricsCollector {
+	if !config.EnableCPUMetrics && !config.EnableMemMetrics && !config.EnableGCMetrics {
+		return nil
+	}
+	rm := NewRuntimeMetricsCollector(config, scope)
+	rm.Start()
+	return rm
+}
+
+// NewRuntimeMetricsCollector creates a new runtime metrics collector.
+func NewRuntimeMetricsCollector(
+	config RuntimeMetricsConfiguration,
+	scope tally.Scope,
+) RuntimeMetricsCollector {
+	return &runtimeCollector{
```

Comment:
We need to tag the host name or application identifier for these metrics to be useful.